File: backend/internal/services/sync_service.go
package services

import (
	"context"
	"time"

	"github.com/google/uuid"
	"github.com/hamishgilbert/notes-app/backend/internal/models"
	"github.com/hamishgilbert/notes-app/backend/internal/repository"
)

const ISO8601Format = "2006-01-02T15:04:05.000Z"

type SyncService struct {
	noteRepo *repository.NoteRepository
}

func NewSyncService(noteRepo *repository.NoteRepository) *SyncService {
	return &SyncService{noteRepo: noteRepo}
}

func (s *SyncService) Sync(ctx context.Context, userID uuid.UUID, req *models.SyncRequest) (*models.SyncResponse, error) {
	// Parse lastSync time
	var lastSync *time.Time
	if req.LastSync != nil && *req.LastSync != "" {
		t, err := time.Parse(ISO8601Format, *req.LastSync)
		if err == nil {
			lastSync = &t
		}
	}

	// Process incoming changes (upsert)
	for _, dto := range req.Changes {
		note, err := s.dtoToNote(dto, userID)
		if err != nil {
			continue // Skip invalid notes
		}
		if err := s.noteRepo.Upsert(ctx, note); err != nil {
			return nil, err
		}
	}

	// Process deletions
	for _, idStr := range req.DeletedIDs {
		id, err := uuid.Parse(idStr)
		if err != nil {
			continue
		}
		// Soft delete - ignore errors for non-existent notes
		_ = s.noteRepo.SoftDelete(ctx, id, userID)
	}

	// Fetch notes updated since lastSync
	notes, err := s.noteRepo.GetAllByUserID(ctx, userID, lastSync)
	if err != nil {
		return nil, err
	}

	// Fetch deleted note IDs since lastSync
	deletedIDs, err := s.noteRepo.GetDeletedSince(ctx, userID, lastSync)
	if err != nil {
		return nil, err
	}

	// Convert to DTOs
	noteDTOs := make([]models.NoteDTO, len(notes))
	for i, note := range notes {
		noteDTOs[i] = s.noteToDTO(&note)
	}

	deletedIDStrings := make([]string, len(deletedIDs))
	for i, id := range deletedIDs {
		deletedIDStrings[i] = id.String()
	}

	return &models.SyncResponse{
		Notes:           noteDTOs,
		DeletedNoteIDs:  deletedIDStrings,
		ServerTimestamp: time.Now().UTC().Format(ISO8601Format),
	}, nil
}

func (s *SyncService) noteToDTO(note *models.Note) models.NoteDTO {
	dto := models.NoteDTO{
		ID:         note.ID.String(),
		Title:      note.Title,
		Content:    note.Content,
		NoteType:   string(note.NoteType),
		IsPinned:   note.IsPinned,
		IsArchived: note.IsArchived,
		SortOrder:  note.SortOrder,
		CreatedAt:  note.CreatedAt.UTC().Format(ISO8601Format),
		UpdatedAt:  note.UpdatedAt.UTC().Format(ISO8601Format),
	}

	if len(note.ChecklistItems) > 0 {
		dto.ChecklistItems = make([]models.ChecklistItemDTO, len(note.ChecklistItems))
		for i, item := range note.ChecklistItems {
			dto.ChecklistItems[i] = models.ChecklistItemDTO{
				ID:          item.ID.String(),
				Text:        item.Text,
				IsCompleted: item.IsCompleted,
				SortOrder:   item.SortOrder,
				CreatedAt:   item.CreatedAt.UTC().Format(ISO8601Format),
				UpdatedAt:   item.UpdatedAt.UTC().Format(ISO8601Format),
			}
		}
	}

	return dto
}

func (s *SyncService) dtoToNote(dto models.NoteDTO, userID uuid.UUID) (*models.Note, error) {
	id, err := uuid.Parse(dto.ID)
	if err != nil {
		return nil, err
	}

	createdAt, err := time.Parse(ISO8601Format, dto.CreatedAt)
	if err != nil {
		createdAt = time.Now()
	}

	updatedAt, err := time.Parse(ISO8601Format, dto.UpdatedAt)
	if err != nil {
		updatedAt = time.Now()
	}

	note := &models.Note{
		ID:         id,
		UserID:     userID,
		Title:      dto.Title,
		Content:    dto.Content,
		NoteType:   models.NoteType(dto.NoteType),
		IsPinned:   dto.IsPinned,
		IsArchived: dto.IsArchived,
		SortOrder:  dto.SortOrder,
		CreatedAt:  createdAt,
		UpdatedAt:  updatedAt,
	}

	// Convert checklist items
	if len(dto.ChecklistItems) > 0 {
		note.ChecklistItems = make([]models.ChecklistItem, len(dto.ChecklistItems))
		for i, itemDTO := range dto.ChecklistItems {
			itemID, err := uuid.Parse(itemDTO.ID)
			if err != nil {
				itemID = uuid.New()
			}

			itemCreatedAt, err := time.Parse(ISO8601Format, itemDTO.CreatedAt)
			if err != nil {
				itemCreatedAt = time.Now()
			}

			itemUpdatedAt, err := time.Parse(ISO8601Format, itemDTO.UpdatedAt)
			if err != nil {
				itemUpdatedAt = time.Now()
			}

			note.ChecklistItems[i] = models.ChecklistItem{
				ID:          itemID,
				NoteID:      note.ID,
				Text:        itemDTO.Text,
				IsCompleted: itemDTO.IsCompleted,
				SortOrder:   itemDTO.SortOrder,
				CreatedAt:   itemCreatedAt,
				UpdatedAt:   itemUpdatedAt,
			}
		}
	}

	return note, nil
}

// NoteToDTO is exported for handlers
func (s *SyncService) NoteToDTO(note *models.Note) models.NoteDTO {
	return s.noteToDTO(note)
}

// DTOToNote is exported for handlers
func (s *SyncService) DTOToNote(dto models.NoteDTO, userID uuid.UUID) (*models.Note, error) {
	return s.dtoToNote(dto, userID)
}

File: ios/Notes/Notes/Services/SyncService.swift
import Foundation
import SwiftData
import Combine

// MARK: - Sync State

enum SyncState: Equatable {
    case idle
    case syncing
    case success(Date)
    case error(String)
}

// MARK: - Sync Service

@Observable
@MainActor
final class SyncService {
    static let shared = SyncService()

    private(set) var syncState: SyncState = .idle
    private(set) var lastSyncDate: Date?

    private var modelContext: ModelContext?
    private var syncTask: Task<Void, Never>?
    private var autoSyncTimer: Timer?

    private let userDefaults = UserDefaults.standard
    private let lastSyncKey = "lastSyncDate"
    private let autoSyncInterval: TimeInterval = 60 // 1 minute
    private let reducedSyncInterval: TimeInterval = 300 // 5 minutes when WebSocket connected

    // WebSocket coordination
    private(set) var isWebSocketConnected = false

    private init() {
        lastSyncDate = userDefaults.object(forKey: lastSyncKey) as? Date
    }

    // MARK: - Configuration

    func configure(modelContext: ModelContext, baseURL: String, authToken: String?) async {
        self.modelContext = modelContext
        await APIClient.shared.configure(baseURL: baseURL, authToken: authToken)
    }

    func setAuthToken(_ token: String?) async {
        await APIClient.shared.setAuthToken(token)
    }

    // MARK: - WebSocket Coordination

    func setWebSocketConnected(_ connected: Bool) {
        isWebSocketConnected = connected

        if connected {
            // WebSocket is active - reduce polling frequency (backup only)
            stopAutoSync()
            autoSyncTimer = Timer.scheduledTimer(
                withTimeInterval: reducedSyncInterval,
                repeats: true
            ) { [weak self] _ in
                Task { @MainActor in
                    await self?.sync()
                }
            }
        } else {
            // WebSocket disconnected - resume normal polling and sync immediately
            stopAutoSync()
            startAutoSync()
            Task {
                await sync()
            }
        }
    }

    // MARK: - Auto Sync

    func startAutoSync() {
        stopAutoSync()
        autoSyncTimer = Timer.scheduledTimer(withTimeInterval: autoSyncInterval, repeats: true) { [weak self] _ in
            Task { @MainActor in
                await self?.sync()
            }
        }
    }

    func stopAutoSync() {
        autoSyncTimer?.invalidate()
        autoSyncTimer = nil
    }

    // MARK: - Manual Sync

    func sync() async {
        guard syncState != .syncing else { return }
        guard let modelContext = modelContext else {
            syncState = .error("Sync service not configured")
            return
        }

        syncState = .syncing

        do {
            // Get pending changes
            let pendingNotes = try fetchPendingNotes(from: modelContext)
            let deletedIDs = try fetchDeletedNoteIDs(from: modelContext)

            // Convert to DTOs
            let noteDTOs = pendingNotes.map { DTOConverter.noteDTO(from: $0) }

            // Sync with server
            let response = try await APIClient.shared.syncNotes(
                changes: noteDTOs,
                deletedIDs: deletedIDs,
                lastSync: lastSyncDate
            )

            // Apply server changes and update serverIDs for pending notes
            try await applyServerChanges(response, to: modelContext, pendingNotes: pendingNotes)

            // Update sync status
            let syncDate = ISO8601DateFormatter().date(from: response.serverTimestamp) ?? Date()
            lastSyncDate = syncDate
            userDefaults.set(syncDate, forKey: lastSyncKey)

            // Mark synced notes (serverID should now be set)
            for note in pendingNotes {
                note.syncStatus = .synced
            }

            // Clear deleted IDs cache
            clearDeletedIDsCache()

            try modelContext.save()

            syncState = .success(syncDate)

        } catch {
            syncState = .error(error.localizedDescription)
        }
    }

    // MARK: - Pending Changes

    private func fetchPendingNotes(from context: ModelContext) throws -> [Note] {
        // Fetch all notes and filter locally since Predicate can't compare enums directly
        let descriptor = FetchDescriptor<Note>()
        let allNotes = try context.fetch(descriptor)
        return allNotes.filter { $0.syncStatus != .synced }
    }

    private func fetchDeletedNoteIDs(from context: ModelContext) throws -> [String] {
        // In a real implementation, you'd store deleted IDs in a separate table
        // For now, return empty array - this would be enhanced with a DeletedNote model
        return userDefaults.stringArray(forKey: "deletedNoteIDs") ?? []
    }

    func markNoteAsDeleted(serverID: String) {
        var deletedIDs = userDefaults.stringArray(forKey: "deletedNoteIDs") ?? []
        if !deletedIDs.contains(serverID) {
            deletedIDs.append(serverID)
            userDefaults.set(deletedIDs, forKey: "deletedNoteIDs")
        }
    }

    private func clearDeletedIDsCache() {
        userDefaults.removeObject(forKey: "deletedNoteIDs")
    }

    // MARK: - Apply Server Changes

    private func applyServerChanges(_ response: SyncResponse, to context: ModelContext, pendingNotes: [Note]) async throws {
        // Fetch all notes with server IDs for comparison
        let allNotesDescriptor = FetchDescriptor<Note>()
        let allNotes = try context.fetch(allNotesDescriptor)

        // Delete notes that were deleted on server
        for deletedID in response.deletedNoteIDs {
            if let noteToDelete = allNotes.first(where: { $0.serverID == deletedID }) {
                context.delete(noteToDelete)
            }
        }

        // Update or create notes from server
        for noteDTO in response.notes {
            try await upsertNote(from: noteDTO, in: context, existingNotes: allNotes, pendingNotes: pendingNotes)
        }
    }

    private func upsertNote(from dto: NoteDTO, in context: ModelContext, existingNotes: [Note], pendingNotes: [Note]) async throws {
        // Try to find existing note by serverID first
        var existingNote = existingNotes.first { $0.serverID == dto.id }

        // If not found by serverID, check if this is a response to a pending note we just uploaded
        // Match by local UUID (pending notes use their local UUID as the ID when sent to server)
        if existingNote == nil {
            existingNote = pendingNotes.first { $0.id.uuidString == dto.id }
        }

        // Also try matching by local UUID in case server preserved it
        if existingNote == nil {
            existingNote = existingNotes.first { $0.id.uuidString == dto.id && $0.serverID == nil }
        }

        let note: Note
        if let existing = existingNote {
            // Check for conflicts (only if note was modified locally after server version)
            let serverDate = ISO8601DateFormatter().date(from: dto.updatedAt) ?? Date()
            if existing.syncStatus == .pending && existing.updatedAt > serverDate {
                // Local changes are newer, mark as conflict but still update serverID
                existing.serverID = dto.id
                existing.syncStatus = .conflict
                return
            }
            note = existing
        } else {
            note = Note()
            context.insert(note)
        }

        // Update note properties
        note.serverID = dto.id
        note.title = dto.title
        note.content = dto.content
        note.noteType = NoteType(rawValue: dto.noteType) ?? .note
        note.isPinned = dto.isPinned
        note.isArchived = dto.isArchived
        note.sortOrder = dto.sortOrder
        note.createdAt = ISO8601DateFormatter().date(from: dto.createdAt) ?? Date()
        note.updatedAt = ISO8601DateFormatter().date(from: dto.updatedAt) ?? Date()
        note.syncStatus = .synced

        // Update checklist items
        if let itemDTOs = dto.checklistItems {
            // Remove existing items
            if let existingItems = note.checklistItems {
                for item in existingItems {
                    context.delete(item)
                }
            }

            // Add new items
            note.checklistItems = itemDTOs.map { itemDTO in
                let item = ChecklistItem(
                    id: UUID(uuidString: itemDTO.id) ?? UUID(),
                    text: itemDTO.text,
                    isCompleted: itemDTO.isCompleted,
                    sortOrder: itemDTO.sortOrder,
                    createdAt: ISO8601DateFormatter().date(from: itemDTO.createdAt) ?? Date(),
                    updatedAt: ISO8601DateFormatter().date(from: itemDTO.updatedAt) ?? Date(),
                    note: note
                )
                context.insert(item)
                return item
            }
        }
    }
}

// MARK: - Sync Status Extension

extension Note {
    var needsSync: Bool {
        syncStatus != .synced
    }

    var hasConflict: Bool {
        syncStatus == .conflict
    }
}
